File: src/org/genyris/interp/ClassloaderFunctions.java
//Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp;

import org.genyris.core.Constants;
import org.genyris.core.Exp;
import org.genyris.core.StrinG;
import org.genyris.exception.GenyrisException;

public class ClassloaderFunctions extends ApplicableFunction {

    public ClassloaderFunctions(Interpreter interp) {
        super(interp, Constants.PREFIX_SYSTEM + "load-class-by-name", true);
    }

    public Exp bindAndExecute(Closure proc, Exp[] arguments,
            Environment envForBindOperations) throws GenyrisException {
        checkArguments(arguments, 1);
        Class[] types = { StrinG.class };
        checkArgumentTypes(types, arguments);
        _interp.loadClassByName(arguments[0].toString());
        return arguments[0];
    }


    public static void bindFunctionsAndMethods(Interpreter interpreter) throws UnboundException, GenyrisException {
        interpreter.bindGlobalProcedure(ClassloaderFunctions.class);
    }

}

File: src/org/genyris/interp/ClassloaderFunctions.kt
//Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp

import org.genyris.core.Constants
import org.genyris.core.Exp
import org.genyris.core.PrefixSymbol
import org.genyris.core.StrinG
import org.genyris.exception.GenyrisException

class ClassloaderFunctions(interp: Interpreter) :
    ApplicableFunction(interp, PrefixSymbol(Constants.PREFIX_SYSTEM, "load-class-by-name", "sys"), true) {
    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(
        proc: Closure?, arguments: Array<Exp?>,
        envForBindOperations: Environment?
    ): Exp? {
        checkArguments(arguments, 1)
        val types = arrayOf<Class<*>?>(StrinG::class.java)
        checkArgumentTypes(types, arguments)
        _interp.loadClassByName(arguments[0].toString())
        return arguments[0]
    }


    companion object {
        @kotlin.Throws(UnboundException::class, GenyrisException::class)
        fun bindFunctionsAndMethods(interpreter: Interpreter) {
            interpreter.bindGlobalProcedureInstance(ClassloaderFunctions(interpreter))
        }
    }
}
